#include "qqcellline.h"
#include "ui_qqcellline.h"
#include "../../src/utils/styleLoader.h"
#include "../../src/utils/appconfig.h"
#include <QNetworkReply>
#include <QNetworkAccessManager>
#include <QMouseEvent>
#include "../../src/utils/utils.h"

QQCellLine::QQCellLine(const QString &avatarUrl,const QString &name, bool isOnLine, QWidget *parent,const int _id)
    : QWidget(parent)
    , ui(new Ui::QQCellLine)
    , m_user_id(_id)
{


    ui->setupUi(this);

    // æ‰“å¼€é¼ æ ‡è·Ÿè¸ªï¼Œæ‰èƒ½å®žæ—¶æŽ¥æ”¶ hover äº‹ä»¶
    setMouseTracking(true);







    // æ‰“å¼€æ ·å¼èƒŒæ™¯æ”¯æŒ
    this->setAttribute(Qt::WA_StyledBackground, true);


    // éªŒè¯èµ„æºæ–‡ä»¶æ˜¯å¦å­˜åœ¨
    QString qssPath = ":/styles/qqcellline.css";
    // åŠ è½½æ ·å¼
    StyleLoader::loadWidgetStyle(this,qssPath);

    //è‡ªåŠ¨ç¼©æ”¾å›¾ç‰‡å†…å®¹
    ui->labHeadeImg->setScaledContents(true);

    setAvatar(avatarUrl);

    ui->labFriendName->setText(name);

    if(isOnLine){
        ui->labStatus->setText(QString("[ åœ¨çº¿ ðŸŸ¢ ]"));
    }else{
        ui->labStatus->setText(QString("[ ç¦»çº¿ ]"));
    }
}

QQCellLine::~QQCellLine()
{
    delete ui;
}

void QQCellLine::setAvatar(const QString &avatarUrl)
{

    QString FileAvatarUrl = AppConfig::instance().imagesDirectory() +"/" +avatarUrl;


    // å¦‚æžœæ˜¯æœ¬åœ°æ–‡ä»¶
    if (QFile::exists(FileAvatarUrl)) {
        QPixmap px(FileAvatarUrl);

        int cornerRadius = 18;  // åœ†è§’åŠå¾„ï¼Œæ ¹æ®ä½ çš„éœ€æ±‚è°ƒæ•´

        QSize s;
        s.setHeight(40);
        s.setWidth(40);


        QPixmap rounded = scaledRoundedPixmap(px,s ,cornerRadius);

        ui->labHeadeImg->setPixmap(rounded);
    } else {
        // ç®€å•ç¤ºä¾‹ï¼šå¦‚æžœæ˜¯ URLï¼Œå¯å¼‚æ­¥åŠ è½½
        QNetworkAccessManager *mgr = new QNetworkAccessManager(this);
        connect(mgr, &QNetworkAccessManager::finished, this, [this,mgr](QNetworkReply *reply){
            if (reply->error() == QNetworkReply::NoError) {
                QByteArray data = reply->readAll();
                QPixmap px;
                px.loadFromData(data);


                QSize s;
                s.setHeight(40);
                s.setWidth(40);


                int cornerRadius = 20;  // åœ†è§’åŠå¾„ï¼Œæ ¹æ®ä½ çš„éœ€æ±‚è°ƒæ•´
                QPixmap rounded = scaledRoundedPixmap(px,s, cornerRadius);

                ui->labHeadeImg->setPixmap(rounded);
            }
            reply->deleteLater();
            mgr->deleteLater();
        });
        mgr->get(QNetworkRequest(QUrl(avatarUrl)));
    }
}

void QQCellLine::setName(const QString &name)
{
    ui->labFriendName->setText(name);
}

void QQCellLine::setOnline(bool online)
{
    if(online){
        ui->labStatus->setText(QString("å½“å‰ç”¨æˆ·çŠ¶æ€ï¼šåœ¨çº¿"));
    }else{
        ui->labStatus->setText(QString("å½“å‰ç”¨æˆ·çŠ¶æ€ï¼šç¦»çº¿"));
    }
}

void QQCellLine::setUserId(int _id)
{
    m_user_id = _id;
}

int QQCellLine::getUserId()
{
    return m_user_id;
}

void QQCellLine::mousePressEvent(QMouseEvent *event)
{
    if (event->button() == Qt::LeftButton) {
        emit clicked();
    }

    QWidget::mousePressEvent(event);
}
